invokeai/app/services/shared/sqlite/sqlite_migrator.py

```diff
@@ -0,0 +1,210 @@
+    :param database: The path to the database file, or ":memory:" for an in-memory database.
+    :param lock: A lock to use when accessing the database.
+    :param logger: The logger to use.
+
+    Migrations should be registered with :meth:`register_migration`. Migrations will be run in
+    order of their version number. If the database is already at the latest version, no migrations
+    will be run.
+    """
+
+    def __init__(self, conn: sqlite3.Connection, database: Path | str, lock: threading.RLock, logger: Logger) -> None:
+        self._logger = logger
+        self._conn = conn
+        self._cursor = self._conn.cursor()
+        self._lock = lock
+        self._database = database
+        self._migrations: set[Migration] = set()
+
+    def register_migration(self, migration: Migration) -> None:
+        """Registers a migration."""
+        if not isinstance(migration.db_version, int) or migration.db_version < 1:
+            raise MigrationVersionError(f"Invalid migration version {migration.db_version}")
+        if any(m.db_version == migration.db_version for m in self._migrations):
+            raise MigrationVersionError(f"Migration version {migration.db_version} already registered")
+        self._migrations.add(migration)
+        self._logger.debug(f"Registered migration {migration.db_version}")
+
+    def run_migrations(self) -> None:
+        """Migrates the database to the latest version."""
+        with self._lock:
+            self._create_version_table()
+            sorted_migrations = sorted(self._migrations, key=lambda m: m.db_version)
+            current_version = self._get_current_version()
+
+            if len(sorted_migrations) == 0:
+                self._logger.debug("No migrations registered")
+                return
+
+            latest_version = sorted_migrations[-1].db_version
+            if current_version == latest_version:
+                self._logger.debug("Database is up to date, no migrations to run")
+                return
+
+            if current_version > latest_version:
+                raise MigrationError(
+                    f"Database version {current_version} is greater than the latest migration version {latest_version}"
+                )
+
+            self._logger.info("Database update needed")
+
+            # Only make a backup if using a file database (not memory)
+            backup_path: Optional[Path] = None
+            if isinstance(self._database, Path):
+                backup_path = self._backup_db(self._database)
+            else:
+                self._logger.info("Using in-memory database, skipping backup")
+
+            for migration in sorted_migrations:
+                try:
+                    self._run_migration(migration)
+                except MigrationError:
+                    if backup_path is not None:
+                        self._logger.error(f" Restoring from {backup_path}")
+                        self._restore_db(backup_path)
```

Comment:
The alternative way to guard against failed migrations is to make a copy of the database as before, make the schema changes _in the new database_, and then when everything completes move the old database away (thereby becoming the backup) and renaming the copy to become the current database.

Reply:
Oh, that sounds much safer